modelforge/potential/neighbors.py

```diff
@@ -0,0 +1,418 @@
+        """
+        Prepares the input tensors for passing to the model.
+
+        This method handles general input manipulation, such as calculating
+        distances and generating the pair list. It also calls the model-specific
+        input preparation.
+
+        Parameters
+        ----------
+        data : NNPInputTuple
+            The input data provided by the dataset, containing atomic numbers,
+            positions, and other necessary information.
+
+        Returns
+        -------
+        PairListOutputs
+            Contains pair indices, distances (d_ij), and displacement vectors (r_ij) for atom pairs within the cutoff.
+        """
+        # ---------------------------
+        # general input manipulation
+        positions = data.positions
+        atomic_subsystem_indices = data.atomic_subsystem_indices
+
+        n = atomic_subsystem_indices.size(0)
+
+        # avoid reinitializing indices if they are already set and haven't changed
+        if self.indices.shape[0] != n:
+            # Generate a range of indices from 0 to n-1
+            self.indices = torch.arange(n, device=atomic_subsystem_indices.device)
+
+            # Create a meshgrid of indices
+            self.i_final_pairs, self.j_final_pairs = torch.meshgrid(
+                self.indices, self.indices, indexing="ij"
+            )
+            # We will only consider unique pairs; for non-unique pairs we can just appropriately copy
+            # the data as it will be faster than extra computations.
+            mask = self.i_final_pairs < self.j_final_pairs
+
+            self.i_final_pairs = self.i_final_pairs[mask]
+            self.j_final_pairs = self.j_final_pairs[mask]
+
+        # calculate r_ij and d_ij
+        r_ij, d_ij = self.displacement_function(
+            positions[self.i_final_pairs],
+            positions[self.j_final_pairs],
+            data.box_vectors,
+        )
+        in_cutoff = (d_ij <= self.cutoff).squeeze()
+        total_pairs = in_cutoff.sum()
+
+        if self.only_unique_pairs:
+            # using this instead of torch.stack to ensure that if we only have a single pair
+            # we don't run into an issue with tensor shapes.
+            # note this will fail if there are no interacting pairs
+
+            pairs = torch.zeros(
+                2, total_pairs, dtype=torch.int64, device=positions.device
+            )
+
+            pairs[0] = self.i_final_pairs[in_cutoff]
+            pairs[1] = self.j_final_pairs[in_cutoff]
+
+            return PairlistData(
+                pair_indices=pairs,
+                d_ij=d_ij[in_cutoff],
+                r_ij=r_ij[in_cutoff],
+            )
+
+        else:
+
+            r_ij_full = torch.zeros(
+                total_pairs * 2, 3, dtype=positions.dtype, device=positions.device
+            )
+            d_ij_full = torch.zeros(
+                total_pairs * 2, 1, dtype=positions.dtype, device=positions.device
+            )
```

Comment:
I am surprised that this logic works.

My understanding is that e.g for a methane, using only unique paris, we have
```
pairs = [0,0,0,0,0,1,1,1,1,2,2,2,3,3,4],
             [1,2,3,4,5,2,3,4,5,3,4,5,4,5,5]
```
which reflects the upper triangle of the squared interaction matrix using index `i` and `j`.
The operation that you are suggesting is effectively:
- generate `d_ij` and `r_ij` for unique pairs
- duplicate and append to generate d_ij for all interactions

resulting in the following interactions
```
pairs = [0,0,0,0,0,1,1,1,1,2,2,2,3,3,4,0,0,0,0,0, ...],
        [1,2,3,4,5,2,3,4,5,3,4,5,4,5,5,1,2,3,4,5,...]
```
which doesn't look right.

But I might be missing something here!

Reply:
We don't just concatenate, we flip indices for the second half. 


so given a 3 atom system with unique pairs with i !=j :

```
pairs = [[0,0,1], 
              [1,2,2]]
```

flipping the indices for the second half becomes:

```
pairs = [[0,0,1,1,2,2], 
             [1,2,2,0,0,1]]
```

Which includes all the relevant non-unique pairs.

Reply:
sorry, my post is a bit missleading: I think what I mean is, that for a given index $i$, the indices $j$ are increasing.